cirq/contrib/acquaintance/strategy.py

```diff
@@ -0,0 +1,135 @@
+# Copyright 2018 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections import defaultdict
+from functools import partial
+from operator import ne
+from typing import Sequence, TYPE_CHECKING, Union
+
+from cirq.circuits import Circuit, Moment, ExpandComposite
+from cirq.extension import Extensions
+from cirq.ops import QubitId, GateOperation
+from cirq.contrib.acquaintance.gates import (
+     SwapNetworkGate, AcquaintanceOpportunityGate, ACQUAINT,
+     op_acquaintance_size)
+from cirq.contrib.acquaintance.permutation import (
+        PermutationGate)
+
+if TYPE_CHECKING:
+    # pylint: disable=unused-import
+    from typing import Type, Dict, List
+    from cirq.ops import Gate
+
+STRATEGY_GATE = Union[AcquaintanceOpportunityGate, PermutationGate]
+
+class AcquaintanceStrategy(Circuit):
```

Comment:
Is it necessary for this class to extend `Circuit`instead of having a circuit? The fact that you pass a circuit into `__init__` strongly hints that the class should be having-a circuit not being-a circuit.

Reply:
What's the benefit of having the circuit as an attribute? Semantically, I think of a strategy as a circuit, albeit with a restricted gate set.

One possibility is defining a `device` whose allowed gates are only `AcquaintanceOpportunityGate` and `PermutationGate` and passing that to the circuit, and then making `AcquaintanceStrategy`s module-level methods.

Reply:
The main reason it's not appropriate to use inheritance in this situation is because your class does not satisfy the [substitution principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle). Child classes may guarantee additional properties (e.g. saying that now a function will never return None), but should not to enforce additional rules (e.g. requiring a function argument to never be None). This class requires that all gates be of a particular type.

Your documentation states that this class is "a special type of circuit that contains only gates of two types". But it is impossible to guarantee this without placing restrictions on what may be passed into e.g. append or insert.

Reply:
It's up to you. If you're going to use circuit methods, it should probably be a circuit. If you're just appending moments, that could just be a list.